src/text-fragment-utils.js

```diff
@@ -539,16 +539,78 @@ const findRangeFromNodeList = (query, range, textNodes) => {
 /**
  * Generates a boundary point pointing to the given text position.
  * @param {Number} index - the text offset indicating the start/end of a
- *     substring of the concatenated text of |textNodes|
+ *     substring of the concatenated, normalized text in |textNodes|
  * @param {Node[]} textNodes - the text Nodes whose contents make up the search
  *     space
  * @param {bool} isEnd - indicates whether the offset is the start or end of the
  *     substring
  * @return {[Node, Number]} - a boundary point suitable for setting as the start
- *     or end of a Range
+ *     or end of a Range, or an empty tuple if it couldn't be computed.
  */
 const getBoundaryPointAtIndex = (index, textNodes, isEnd) => {
-  // Iterate over lengths
+  let counted = 0;
+  let normalizedData;
+  for (let i = 0; i < textNodes.length; i++) {
+    const node = textNodes[i];
+    if (!normalizedData) normalizedData = normalizeString(node.data);
+    let nodeEnd = counted + normalizedData.length;
+    if (isEnd) nodeEnd += 1;
+    if (nodeEnd > index) {
+      // |index| falls within this node, but we need to turn the offset in the
+      // normalized data into an offset in the real node data.
+      const normalizedOffset = index - counted;
+      let denormalizedOffset = Math.min(index - counted, node.data.length);
+
+      // Walk through the string until denormalizedOffset produces a substring
+      // that corresponds to the target from the normalized data.
+      const targetSubstring = isEnd
+        ? normalizedData.substring(0, normalizedOffset)
+        : normalizedData.substring(normalizedOffset);
+
+      let candidateSubstring = isEnd
+        ? normalizeString(node.data.substring(0, denormalizedOffset))
+        : normalizeString(node.data.substring(denormalizedOffset));
+
+      // We will either lengthen or shrink the candidate string to approach the
+      // length of the target string. If we're looking for the start, adding 1
+      // makes the candidate shorter; if we're looking for the end, it makes the
+      // candidate longer.
+      const direction =
+        (isEnd ? -1 : 1) *
+        (targetSubstring.length > candidateSubstring.length ? -1 : 1);
+
+      while (
+        denormalizedOffset >= 0 &&
+        denormalizedOffset <= node.data.length
+      ) {
+        if (candidateSubstring.length === targetSubstring.length)
+          return [node, denormalizedOffset];
```

Comment:
nit: For readability, I personally prefer returning objects with named values instead of an array with two values, like:
`return {node, offset:denormalizedOffset};`

And in the failure case, returning undefined rather than an empty array (allowing for falsy checks).